src/cli/cli_tcp.hpp

```diff
@@ -0,0 +1,160 @@
+    static void HandleTcpDisconnectedCallback(otTcpEndpoint *aEndpoint, otTcpDisconnectedReason aReason);
+    static otTcpIncomingConnectionAction HandleTcpAcceptReadyCallback(otTcpListener *   aListener,
+                                                                      const otSockAddr *aPeer,
+                                                                      otTcpEndpoint **  aAcceptInto);
+    static void                          HandleTcpAcceptDoneCallback(otTcpListener *   aListener,
+                                                                     otTcpEndpoint *   aEndpoint,
+                                                                     const otSockAddr *aPeer);
+
+    void                          HandleTcpEstablished(otTcpEndpoint *aEndpoint);
+    void                          HandleTcpSendDone(otTcpEndpoint *aEndpoint, otLinkedBuffer *aData);
+    void                          HandleTcpReceiveAvailable(otTcpEndpoint *aEndpoint,
+                                                            size_t         aBytesAvailable,
+                                                            bool           aEndOfStream,
+                                                            size_t         aBytesRemaining);
+    void                          HandleTcpDisconnected(otTcpEndpoint *aEndpoint, otTcpDisconnectedReason aReason);
+    otTcpIncomingConnectionAction HandleTcpAcceptReady(otTcpListener *   aListener,
+                                                       const otSockAddr *aPeer,
+                                                       otTcpEndpoint **  aAcceptInto);
+    void HandleTcpAcceptDone(otTcpListener *aListener, otTcpEndpoint *aEndpoint, const otSockAddr *aPeer);
+
+    static constexpr Command sCommands[] = {
+        {"abort", &TcpExample::ProcessAbort},
+        {"benchmark", &TcpExample::ProcessBenchmark},
+        {"bind", &TcpExample::ProcessBind},
+        {"connect", &TcpExample::ProcessConnect},
+        {"deinit", &TcpExample::ProcessDeinit},
+        {"help", &TcpExample::ProcessHelp},
+        {"init", &TcpExample::ProcessInit},
+        {"listen", &TcpExample::ProcessListen},
+        {"send", &TcpExample::ProcessSend},
+        {"sendend", &TcpExample::ProcessSendEnd},
+        {"stoplistening", &TcpExample::ProcessStopListening},
+    };
+
+    static_assert(Utils::LookupTable::IsSorted(sCommands), "Command Table is not sorted");
+
+    Interpreter &mInterpreter;
+
+    otTcpEndpoint mEndpoint;
+    otTcpListener mListener;
+
+    bool mEndpointInitialized;
+    bool mListenerInitialized;
+    bool mEndpointConnected;
+    bool mSendBusy;
+
+    otLinkedBuffer mSendLink;
+    uint8_t        mSendBuffer[OPENTHREAD_CONFIG_CLI_MAX_LINE_LENGTH];
+    uint8_t        mReceiveBuffer[OPENTHREAD_CONFIG_CLI_TCP_RECEIVE_BUFFER_SIZE];
+
+    otLinkedBuffer mBenchmarkLinks[(sizeof(mReceiveBuffer) + sizeof(mSendBuffer) - 1) / sizeof(mSendBuffer)];
```

Comment:
I'm happy to add a separate configuration flag for the TCP CLI. But, I guess I'm not 100% clear on why it's needed? I had assumed that the CLI tool wouldn't actually be used in production, so it's memory footprint doesn't matter for real applications. Am I mistaken?

Reply:
I think it's logical for a user program on a SoC device to use TCP and CLI, but don't need TCP CLI ?

Reply:
I had thought of the CLI as being a standalone application. I didn't realize it could be part of another user program. Can you please elaborate?

I'll add the flag, but I'd like to understand this use case more. :)

Reply:
Shall I add the flag to cli_config.h?

Reply:
For example, otbr-agent is a Border Router program, but has all CLI commands integrated. 

I think CLI commands are so convenient that users may just add CLI library to whatever application they are building.

Reply:
`cli_config.h` works for me.